sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDropColumns.scala

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, Project, UnresolvedDropColumns}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.DROP_COLUMNS
+import org.apache.spark.sql.connector.catalog.CatalogManager
+
+/**
+ * A rule that rewrites DropColumns to Project.
+ */
+class ResolveDropColumns(val catalogManager: CatalogManager)
+  extends Rule[LogicalPlan] with ColumnResolutionHelper  {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+    _.containsPattern(DROP_COLUMNS)) {
+    case d: UnresolvedDropColumns if d.childrenResolved =>
+      val dropped = d.dropList.map {
+        case u: UnresolvedAttribute => resolveExpressionByPlanChildren(u, d.child)
+        case e => e
+      }
+      val remaining = d.child.output.filterNot(attr => dropped.exists(_.semanticEquals(attr)))
```

Comment:
what if people doing `df.drop(df.c1 + df.c2)`? will we fail or ignore it?

Reply:
we just ignore it

Reply:
```
scala> val df1 = spark.range(0, 10)
df1: org.apache.spark.sql.Dataset[Long] = [id: bigint]

scala> val df2 = spark.range(0, 5)
df2: org.apache.spark.sql.Dataset[Long] = [id: bigint]

scala> df1.drop(df1.col("id") + df1.col("id"))
res4: org.apache.spark.sql.DataFrame = [id: bigint]

scala> df1.drop(df1.col("id") + df2.col("id"))
res5: org.apache.spark.sql.DataFrame = [id: bigint]

scala> df1.drop(df2.col("id"))
res6: org.apache.spark.sql.DataFrame = [id: bigint]
```

we should also ignore incorrect columns